src/library/serato/seratofeature.cpp

```diff
@@ -0,0 +1,1043 @@
+        case FieldId::SampleRate:
+            track.samplerate = parseText(data, fieldSize);
+            break;
+        case FieldId::Bpm: {
+            bool ok;
+            double bpm = parseText(data, fieldSize).toDouble(&ok);
+            if (ok) {
+                track.bpm = bpm;
+            }
+            break;
+        }
+        case FieldId::Comment:
+            track.comment = parseText(data, fieldSize);
+            break;
+        case FieldId::Grouping:
+            track.grouping = parseText(data, fieldSize);
+            break;
+        case FieldId::Label:
+            track.label = parseText(data, fieldSize);
+            break;
+        case FieldId::Year: {
+            bool ok;
+            int year = parseText(data, fieldSize).toInt(&ok);
+            if (ok) {
+                track.year = year;
+            }
+            break;
+        }
+        case FieldId::Key:
+            track.key = parseText(data, fieldSize);
+            break;
+        case FieldId::BeatgridLocked:
+            track.beatgridlocked = parseBoolean(data);
+            break;
+        case FieldId::Missing:
+            track.missing = parseBoolean(data);
+            break;
+        case FieldId::FileTime:
+            track.filetime = parseUInt32(data);
+            break;
+        case FieldId::DateAdded:
+            track.datetimeadded = parseUInt32(data);
+            break;
+        case FieldId::DateAddedText:
+            // Ignore this field, but do not print a debug message
+            break;
+        default:
+            qDebug() << "Ignoring unknown field "
+                     << fieldName
+                     << " ("
+                     << fieldSize
+                     << " bytes).";
+        }
+
+        headerData = buffer.read(8);
+    }
+
+    if (headerData.length() != 0) {
+        qWarning() << "Found "
+                   << headerData.length()
+                   << " extra bytes at end of track definition.";
+        return false;
+    }
+
+    if (track.location.isEmpty()) {
+        qWarning() << "Found track with empty location field.";
+        return false;
+    }
+
+    return true;
+}
+
+inline QString parseCrateTrack(QIODevice& buffer) {
+    QString location;
+    QByteArray headerData = buffer.read(8);
+    while (headerData.length() == 8) {
+        QString fieldName = QString(headerData.mid(0, 4));
+        quint32 fieldId = parseUInt32(headerData.mid(0, 4));
+        quint32 fieldSize = parseUInt32(headerData.mid(4, 8));
+
+        // Read field data
+        QByteArray data = buffer.read(fieldSize);
+        if (static_cast<quint32>(data.length()) != fieldSize) {
+            qWarning() << "Failed to read "
+                       << fieldSize
+                       << " bytes for "
+                       << fieldName
+                       << " field.";
+            return QString();
+        }
+
+        // Parse field data
+        switch (static_cast<FieldId>(fieldId)) {
```

Comment:
I think we should avoid casting an integer to an enum. 
can we cast the case labels to quint32 instead?

Reply:
Can you elaborate why? Casting in every case label is quite ugly. And I specified an explicit underlying type for the enum, so it should be safe: https://stackoverflow.com/a/55404179

Reply:
Oh, I did not know this. So it is fine in this case. I guess there are some other places in the code where we can use this. Is a double typed enum possible?